File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/external/global-functions/testInvokeExternalGlobalFunction.ts
import { concat, u32ToBuffer, IExternalWrapInstance, BaseTypeSerialization } from '@polywrap/reim-wrap-js';
import { WrapModule } from '../module/WrapModule';
import { WrapManifest } from '../../WrapManifest';
import { ExternalResource } from '../../dt/ExternalResource';


export function testInvokeExternalGlobalFunction(
  arg: string,
): Promise<string> {
  return testInvokeExternalGlobalFunctionFromInstance(
    WrapModule.wrapInstance,
    arg,
  );
};

export const create = (instance: IExternalWrapInstance) => {
  return (
    arg: string,
  ): Promise<string> => {
    return testInvokeExternalGlobalFunctionFromInstance(
      instance, 
      arg,
    );
  };
};

export const testInvokeExternalGlobalFunctionFromInstance = async (
  instance: IExternalWrapInstance | null, 
  arg: string,
): Promise<string> => {
  if (instance == null) {
    throw new Error("connect() or import() must be called before using this module");
  }

  const args = new TestInvokeExternalGlobalFunctionArgs(
    arg,
  );

  const buffer = concat([
    u32ToBuffer(WrapManifest.External.GlobalFunction.TestInvokeExternalGlobalFunction),
    TestInvokeExternalGlobalFunctionArgsWrapped.serialize(args),
  ]);

  const result = await instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer);

  
  return BaseTypeSerialization.deserialize<string>(result);
}

export class TestInvokeExternalGlobalFunctionArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static serialize(value: TestInvokeExternalGlobalFunctionArgs): Uint8Array {
    return new TextEncoder().encode(
      JSON.stringify(
        TestInvokeExternalGlobalFunctionArgsWrapped.mapToSerializable(value)
      )
    );
  }

  static mapToSerializable(value: TestInvokeExternalGlobalFunctionArgs): TestInvokeExternalGlobalFunctionArgsWrapped {
    return new TestInvokeExternalGlobalFunctionArgsWrapped(
            
      value.arg,
      
    );
  }
}

export class TestInvokeExternalGlobalFunctionArgs {
  constructor(
    public arg: string,
  ) {
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/external/global-functions/testInvokeExternalGlobalFunction.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use dt::{ ExternalResource };
use external_module::{ WrapModule };
use wrapped::{ testInvokeExternalGlobalFunctionWrapped };


pub async fn testInvokeExternalGlobalFunction(
  arg: string,
) -> string {
  testInvokeExternalGlobalFunctionFromInstance(
    WrapModule.wrapInstance,
    arg,
  )
};

pub fn create(instance: dyn ExternalWrapModule) {
  |
    arg: string,
  | {
    testInvokeExternalGlobalFunctionFromInstance(
      instance, 
      arg,
    )
  }
}

pub async fn testInvokeExternalGlobalFunctionFromInstance (
  instance: Option<ExternalWrapModule>, 
  arg: string,
) -> string {
  if instance.is_none() {
    panic!("connect() or import() must be called before using this module");
  }

  let args = TestInvokeExternalGlobalFunctionArgs.new(
    arg,
  );

  let buffer = [
    WrapManifest.External.GlobalFunction.TestInvokeExternalGlobalFunction.to_be_bytes(),
    TestInvokeExternalGlobalFunctionArgsWrapped.serialize(args),
  ].concat();

  let result = instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer).await;

  
  BaseTypeSerialization.deserialize<string>(result)
}

struct TestInvokeExternalGlobalFunctionArgsWrapped {
    pub arg: string,
}
impl TestInvokeExternalGlobalFunctionArgsWrapped {
    pub fn new(
      arg: string,
    ) -> TestInvokeExternalGlobalFunctionArgsWrapped {
      TestInvokeExternalGlobalFunctionArgsWrapped {
        arg,
      }
    }

    pub fn serialize(value: TestInvokeExternalGlobalFunctionArgs) -> Vec<u8> {
        json!(
            TestInvokeExternalGlobalFunctionArgsWrapped.mapToSerializable(value)
        ).as_bytes()
    }

    pub fn mapToSerializable(value: TestInvokeExternalGlobalFunctionArgs) -> TestInvokeExternalGlobalFunctionArgsWrapped {
        TestInvokeExternalGlobalFunctionArgsWrapped.new(
                        
            value.arg,
            
        )
    }
}

struct TestInvokeExternalGlobalFunctionArgs {
    pub arg: string,
}

impl TestInvokeExternalGlobalFunctionArgs {
  pub fn new(
    arg: string,
  ) -> TestInvokeExternalGlobalFunctionArgs {
    TestInvokeExternalGlobalFunctionArgs.new (
      arg,
    )
  }
}
